File: Data Structures & Algorithms/validate-parentheses/submission-5.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        for c in s:
            match c:
                case '(':
                    stack.append(c)
                case '{':
                    stack.append(c)
                case '[':
                    stack.append(c)
                case ')':
                    if len(stack) == 0 or not stack.pop() == '(':
                        return False
                case '}':
                    if len(stack) == 0 or not stack.pop() == '{':
                        return False
                case ']':
                    if len(stack) == 0 or not stack.pop() == '[':
                        return False
        return len(stack) == 0

File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-5.swift
class Solution {
    func largestRectangleArea(_ heights: [Int]) -> Int {
        var stack: [(Int, Int)] = []
        var maxArea = -1

        // O(n)
        for (i, val) in heights.enumerated() {
            var lastPoppedI = -1
            
            while !stack.isEmpty && stack.last!.1 > val {
                let popped = stack.popLast()!
                lastPoppedI = popped.0
                let area = (i - popped.0) * popped.1

                if area > maxArea {
                    maxArea = area
                }
            }

            if lastPoppedI == -1 {
                stack.append((i, val))
            } else {
                stack.append((lastPoppedI, val))
            }
        }

        // O(n)
        print(stack)
        while !stack.isEmpty {
            let popped = stack.popLast()!
            let area = (heights.count - popped.0) * popped.1
            print(area)

            if area > maxArea {
                maxArea = area
            }
        }

        return maxArea
    }
}
